_posts/2019-01-24-how-to-keep-safe-as-a-client.md

```diff
@@ -0,0 +1,129 @@
+- Threshold: for example, there are 100 success requests within 5 seconds.
+- Percentage: for example, there are 95% success requests within 10 seconds.
+
+The standard also contains a time window, threshold, and percentage. But there is a slight difference: most of the time, the system will stay unavailable for a moment, So we can check in fixed times(30 seconds for example).
+
+```python
+successCount = 0;
+isHalfOpen = true;
+
+if(success){
+  if(isHalfOpen){
+    successCount ++;
+    if(successCount == threshold){
+      isOpenCircuitBreaker = false;
+    }
+  }
+  return success;
+}else{
+  errorCount ++;
+  if(errorCount == threshold){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+Since the reconnection can still fail, we cannot use all servers to try to reconnect in a large system.. We can take these measures:
+- 1. Use a few servers to check
+- 2. Add a health check interface to check system conditions: CPU, IO etc.
+
+### recover
+
+Once the system is available, we need to recover.
+
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSUGPwicRFU6xJeNPxzKOWEhlM3yKSVdEfk1nusm5FfnRm1WyYkOVybBg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
+This is the general process of Circuit Breaker, frameworks like hystrix, resilience4j are all had such process flow.
+
+## Best practice for Circuit Breaker
+
+### What kind of situation require Circuit Breaker
+
+Circuit Breaker is not suitable for every situation. There are some conditions we can consider for using Circuit Breaker:
+- If the target system is a shared service: when other systems abuse this shared service, our client can be affected.
+- The target system is not in regular use, sometimes the system may be under heavy load, while other times it may be rarely invoked, but we need to maintain stable service during sudden spikes in load.
+
+### Things to watch out for when implementing a circuit breaker
+
+Circuit Breaker is not perfect, we need to pay attention to two things.
+- Firstly, if we rely on a system that has many replicas, we need to keep in mind that the exception in a partial node does not mean the exception of the whole system.
+- Secondly, Circuit Breaker should be the last resort, we can to use "fall back" or "rate limit" prior to Circuit Breaker. Because partial is better than none, although some users may find out that the service is broken, we can try to keep our system working and serve other users. For example, by abandoning non-core services, or by giving friendly hints, we can keep our core services working.
+
+## Conclusion
+
+Circuit Breaker can be implemented before the actual function or after the actual function. It meets the AOP(aspect-oriented programming) concept, so generally, most circuit breaker frameworks use AOP(aspect-oriented programming) as an implemented strategy.
```

Comment:
You can just add "(aspect-oriented programming)" after the first use of AOP. For later uses people will know what it means already. So I think it's better to remove the second "(aspect-oriented programming).